File: 14-longest-common-prefix/solution.test.js
import { test, expect } from '@jest/globals'
import longestCommonPrefix from './solution.js'

test('longestCommonPrefix should find common prefixes in basic cases', () => {
  expect(longestCommonPrefix(['flower', 'flow', 'flight'])).toBe('fl')
  expect(longestCommonPrefix(['dog', 'racecar', 'car'])).toBe('')
  expect(longestCommonPrefix(['apple', 'app', 'application'])).toBe('app')
  expect(longestCommonPrefix(['interstellar', 'interview', 'interest']))
    .toBe('inter')
  expect(longestCommonPrefix(['programming', 'progress', 'progressive']))
    .toBe('progr')
})

test('longestCommonPrefix should handle edge cases with empty strings', () => {
  expect(longestCommonPrefix(['', 'flow', 'flight'])).toBe('')
  expect(longestCommonPrefix(['flower', '', 'flight'])).toBe('')
  expect(longestCommonPrefix(['flower', 'flow', ''])).toBe('')
  expect(longestCommonPrefix(['', '', ''])).toBe('')
})

test('longestCommonPrefix should handle single string arrays', () => {
  expect(longestCommonPrefix(['single'])).toBe('single')
  expect(longestCommonPrefix([''])).toBe('')
})

test('longestCommonPrefix should handle identical strings', () => {
  expect(longestCommonPrefix(['same', 'same', 'same'])).toBe('same')
  expect(longestCommonPrefix(['a', 'a', 'a'])).toBe('a')
})

test('longestCommonPrefix should handle varying string lengths', () => {
  expect(longestCommonPrefix(['a', 'ab', 'abc', 'abcd'])).toBe('a')
  expect(longestCommonPrefix(['abcd', 'abc', 'ab', 'a'])).toBe('a')
  expect(longestCommonPrefix(['long', 'longer', 'longest'])).toBe('long')
})

test('longestCommonPrefix should handle special characters', () => {
  expect(longestCommonPrefix(['123abc', '123def', '123ghi'])).toBe('123')
  expect(longestCommonPrefix(['!@#', '!@#$', '!@#%'])).toBe('!@#')
  expect(longestCommonPrefix(['app-store', 'app-le', 'app-rentice'])).toBe('app-')
})

test('longestCommonPrefix should handle case sensitivity', () => {
  expect(longestCommonPrefix(['Apple', 'app', 'application'])).toBe('')
  expect(longestCommonPrefix(['HELLO', 'HELP', 'HELMET'])).toBe('HEL')
  expect(longestCommonPrefix(['JavaScript', 'javascript'])).toBe('')
})

test('longestCommonPrefix should handle no common prefix', () => {
  expect(longestCommonPrefix(['abc', 'def', 'ghi'])).toBe('')
  expect(longestCommonPrefix(['cat', 'dog', 'rabbit'])).toBe('')
  expect(longestCommonPrefix(['12345', '54321'])).toBe('')
})

test('longestCommonPrefix should handle Unicode characters', () => {
  expect(longestCommonPrefix(['café', 'café-au-lait'])).toBe('café')
  expect(longestCommonPrefix(['résumé', 'résultat'])).toBe('résu')
  expect(longestCommonPrefix(['🍎apple', '🍎app', '🍎application'])).toBe('🍎app')
})

test('longestCommonPrefix should handle large input strings', () => {
  const lng_pre = 'a'.repeat(1000)
  const str1 = lng_pre + 'xyz'
  const str2 = lng_pre + '123'
  const str3 = lng_pre + 'abc'
  expect(longestCommonPrefix([str1, str2, str3])).toBe(lng_pre)
})

File: 14-longest-common-prefix/solution_test.go
package solutionLCP

import (
	"strings"
	"testing"
)

type TestHarness struct {
	strs   []string
	expect string
}

const err_strf = "LongestCommonPrefix(%v) = [got: %q], [expect: %q]"

func checkHarness(t *testing.T, testCases []TestHarness, suite string) {
	t.Run(suite, func(t *testing.T) {
		for _, tt := range testCases {
			got := LongestCommonPrefix(tt.strs)
			if got != tt.expect {
				t.Errorf(err_strf, tt.strs, got, tt.expect)
			}
		}
	})
}

func TestLongestCommonPrefix(t *testing.T) {
	basic := []TestHarness{
		{strs: []string{"flower", "flow", "flight"}, expect: "fl"},
		{strs: []string{"dog", "racecar", "car"}, expect: ""},
		{strs: []string{"apple", "app", "application"}, expect: "app"},
		{strs: []string{"interstellar", "interview", "interest"}, expect: "inter"},
		{strs: []string{"programming", "progress", "progressive"}, expect: "progr"},
	}
	single := []TestHarness{
		{strs: []string{"single"}, expect: "single"},
		{strs: []string{""}, expect: ""},
	}
	dupes := []TestHarness{
		{strs: []string{"same", "same", "same"}, expect: "same"},
		{strs: []string{"a", "a", "a"}, expect: "a"},
	}
	varied := []TestHarness{
		{strs: []string{"a", "ab", "abc", "abcd"}, expect: "a"},
		{strs: []string{"abcd", "abc", "ab", "a"}, expect: "a"},
		{strs: []string{"long", "longer", "longest"}, expect: "long"},
	}
	tests_sane := append(basic, single...)
	tests_sane = append(tests_sane, dupes...)
	tests_sane = append(tests_sane, varied...)
	suite_sane := "Basic inputs, single elem array, dupes in array, varied length"
	checkHarness(t, tests_sane, suite_sane)

	empty := []TestHarness{
		{strs: []string{"", "flow", "flight"}, expect: ""},
		{strs: []string{"flower", "", "flight"}, expect: ""},
		{strs: []string{"flower", "flow", ""}, expect: ""},
		{strs: []string{"", "", ""}, expect: ""},
	}
	special := []TestHarness{
		{strs: []string{"123abc", "123def", "123ghi"}, expect: "123"},
		{strs: []string{"!@#", "!@#$", "!@#%"}, expect: "!@#"},
		{strs: []string{"app-store", "app-le", "app-rentice"}, expect: "app-"},
	}
	ulcase := []TestHarness{
		{strs: []string{"Apple", "app", "application"}, expect: ""},
		{strs: []string{"HELLO", "HELP", "HELMET"}, expect: "HEL"},
		{strs: []string{"JavaScript", "javascript"}, expect: ""},
	}
	no_lcp := []TestHarness{
		{strs: []string{"abc", "def", "ghi"}, expect: ""},
		{strs: []string{"cat", "dog", "rabbit"}, expect: ""},
		{strs: []string{"12345", "54321"}, expect: ""},
	}
	unicode := []TestHarness{
		{strs: []string{"café", "café-au-lait"}, expect: "café"},
		{strs: []string{"résumé", "résultat"}, expect: "résu"},
		{strs: []string{"🍎apple", "🍎app", "🍎application"}, expect: "🍎app"},
	}
	tests_edge := append(empty, special...)
	tests_edge = append(tests_edge, ulcase...)
	tests_edge = append(tests_edge, no_lcp...)
	tests_edge = append(tests_edge, unicode...)
	suite_edge := "Empty strings, special/unicode chars, case sensitivity, no LCP"
	checkHarness(t, tests_edge, suite_edge)

	// Large input strings
	t.Run("large input strings", func(t *testing.T) {
		prefix, strs := buildLrgStrs()
		got := LongestCommonPrefix(strs)
		if got != prefix {
			t.Errorf(err_strf, strs, got, prefix)
		}
	})
}

// Benchmark
func BenchmarkLongestCommonPrefix(b *testing.B) {
	strs := []string{"flower", "flow", "flight"}

	for b.Loop() {
		LongestCommonPrefix(strs)
	}

	_, strs_lrg := buildLrgStrs()

	for b.Loop() {
		LongestCommonPrefix(strs_lrg)
	}
}

func buildLrgStrs() (string, []string) {
	prefix := strings.Repeat("a", 1000)

	strs := []string{
		prefix + "xyz",
		prefix + "123",
		prefix + "abc",
	}

	return prefix, strs
}
